### ISSUE TITLE
ACME certificate with ECDSA algorithm

### ISSUE THREAD

**Issue by GeckoSplinter** [containous/traefik#2940]
### Do you want to request a *feature* or report a *bug*?

Feature

### What did you do?

Traefik configuration is working well with RSA ACME certificates  and Traefik is working well with custom ECDSA certificates.

### What did you expect to see?

A configuration option to request ACME certificates with ECDSA algorithm instead of RSA.
An options to setup in [acme] part of the toml file could be great.

### What did you see instead?

Certificates generated with ACME are using RSA algorithm.

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
Version:      v1.5.2
Codename:     cancoillotte
Go version:   go1.9.4
Built:        2018-02-12_10:56:31AM
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

```toml
debug = false
checkNewVersion = true
sendAnonymousUsage = true
logLevel = "ERROR"
defaultEntryPoints = ["https","http"]

[entryPoints]
  [entryPoints.http]
  address = ":80"
    [entryPoints.http.redirect]
    entryPoint = "https"
  [entryPoints.https]
  address = ":443"
    [entryPoints.https.tls]
    minVersion = "VersionTLS12"
    cipherSuites = [
      "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256",
      "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384",
      "TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305",
      "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256",
      "TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384",
      "TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256",
      "TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256"
    ]
    # default certificate with ecdsa (internal CA)
    [[entryPoints.https.tls.certificates]]
      certFile = "/certs/certificate.pem"
      keyFile = "/certs/key.pem"



[retry]

[docker]
endpoint = "unix:///var/run/docker.sock"
domain = "----"
watch = true
exposedbydefault = false

[acme]
email = "----"
storage = "acme.json"
entryPoint = "https"
OnHostRule = true
dnsProvider = "----"
delayDontCheckDNS = 20

[api]
 entryPoint = "traefik"
 dashboard = true
 debug = false
  [api.auth.basic]
  users = ["----"]
  [api.statistics]
   recentErrors = 10

[metrics]
 [metrics.prometheus]
  entryPoint = "traefik"
  buckets = [0.1,0.3,1.2,5.0]
```

### Goals

Trying to get A+ grade on SSLLabs with traefik

#1814

Let's encrypt handles ECDSA signing since February 10, 2016
Let's encrypt will provide ECDSA Root and Intermediates => ETA: Q3 2018 
https://letsencrypt.org/upcoming-features/

Results using https://github.com/mozilla/tls-observatory

```
--- Analyzers ---
* Mozilla evaluation: intermediate
  - for modern level: remove ciphersuites ECDHE-RSA-AES128-SHA, ECDHE-RSA-AES256-SHA
  - for modern level: consider adding ciphers ECDHE-ECDSA-AES256-GCM-SHA384, ECDHE-ECDSA-CHACHA20-POLY1305, ECDHE-RSA-CHACHA20-POLY1305, ECDHE-ECDSA-AES128-GCM-SHA256, ECDHE-ECDSA-AES256-SHA384, ECDHE-RSA-AES256-SHA384, ECDHE-ECDSA-AES128-SHA256, ECDHE-RSA-AES128-SHA256
  - for modern level: consider enabling OCSP stapling
  - for modern level: use a certificate of type ecdsa, not RSA
  - oldest clients:
* Grade: A (93/100)
* SSLLabs Client Support: showing oldest known clients
  - Firefox 27, Chrome 30, Edge 12, IE 11, Safari 5, Opera 17, Android 4.4.2, OpenSSL 1.0.1h, Java 8b132
```
Maybe to link with #811 Enable Perfect Forward Secrecy

**Comment by navinSing** [comment]
getting same issue .. 
is it fixed ?
LOGS : 
Failed to build ACME client: private key was nil\nLet's Encrypt functionality will be limited until Traefik is restarted.

**Comment by nmengin** [comment]
Hello @navinSing 

For the moment, the issue is still opened.
The enhancement has not been created.

**Comment by schemen** [comment]
Hi!
Do we have any update on this issue?

With a clean setup as per cluster guide I cannot seem to be able to get it to run.
Data seems to get saved successfully into consul though.

Docker-compose file:
```
version: "3.4"
services:
  traefik_init:
    image: traefik:tetedemoine
    command:
      - "storeconfig"
      - "--api"
      - "--entrypoints=Name:http Address::80 Redirect.EntryPoint:https"
      - "--entrypoints=Name:https Address::443 TLS"
      - "--defaultentrypoints=http,https"
      - "--acme"
      - "--acme.storage=traefik/acme/account"
      - "--acme.entryPoint=https"
      - "--acme.httpChallenge.entryPoint=http"
      - "--acme.onHostRule=true"
      - "--acme.onDemand=false"
      - "--acme.caServer=https://acme-staging.api.letsencrypt.org/directory"
      - "--acme.email=#####"
      - "--docker"
      - "--docker.swarmMode"
      - "--docker.domain=#####"
      - "--docker.watch"
      - "--consul"
      - "--consul.endpoint=consul:8500"
      - "--consul.prefix=traefik"
    networks:
      - traefik
    deploy:
      restart_policy:
        condition: on-failure
    depends_on:
      - consul
  traefik:
    image: traefik:tetedemoine
    depends_on:
      - traefik_init
      - consul
    command:
      - "--consul"
      - "--consul.endpoint=consul:8500"
      - "--consul.prefix=traefik"
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    networks:
      - public 
      - traefik
    ports:
      - target: 80
        published: 80
        mode: host
      - target: 443
        published: 443
        mode: host
      - target: 8080
        published: 8080
        mode: host
    deploy:
      mode: global
      placement:
        constraints:
          - node.role == manager
      update_config:
        parallelism: 1
        delay: 10s
      restart_policy:
        condition: on-failure
  consul:
    image: consul
    command: agent -server -bootstrap-expect=1
    volumes:
      - /var/data/traefik_devel/consul:/consul/data
    environment:
      - CONSUL_LOCAL_CONFIG={"datacenter":"falkenstein","server":true}
      - CONSUL_BIND_INTERFACE=eth0
      - CONSUL_CLIENT_INTERFACE=eth0
    deploy:
      replicas: 1
      placement:
        constraints:
          - node.role == manager
      restart_policy:
        condition: on-failure
    networks:
      - traefik

networks:
  public:
    driver: overlay
  traefik:
    driver: overlay

``` 


```
time="2018-05-02T16:06:37Z" level=error msg="Error building ACME client &{Email: Registration:<nil> PrivateKey:[] DomainsCertificate:{Certs:[] lock:{w:{state:0 sema:0} writerSem:0 readerSem:0 readerCount:0 readerWait:0}} ChallengeCerts:map[] HTTPChallenge:map[]}: private key was nil"
time="2018-05-02T16:06:37Z" level=error msg="Cannot unmarshall private key []"```

**Comment by djeeg** [comment]
Have the same issue right now, website is blocked because cert is invalid.

```
2018-05-03T16:13:53.508871214Z {"level":"error","msg":"Cannot unmarshall private key []","time":"2018-05-03T16:13:53Z"}
2018-05-03T16:13:53.514463168Z {"level":"error","msg":"Error building ACME client \u0026{Email: Registration:\u003cnil\u003e PrivateKey:[] DomainsCertificate:{Certs:[] lock:{w:{state:0 sema:0} writerSem:0 readerSem:0 readerCount:0 readerWait:0}} ChallengeCerts:map[] HTTPChallenge:map[]}: private key was nil","time":"2018-05-03T16:13:53Z"}
```

Have tried completely tearing down and rebuilding traefik and etcd, it didnt help. Same error. (though I dont have custom CipherSuites)

**Comment by vlad-tuznichenko** [comment]
Hello!
Do we have any update on this issue?

**Comment by mmatur** [comment]
I will propose a PR this week or next week guys!! Sorry for the delay

**Comment by patrick-motard** [comment]
@mmatur Does this relate to the documentation example [here](https://docs.traefik.io/user-guide/cluster-docker-consul/#migrate-configuration-to-consul)? I'm running into the exact same error after following the example.

**Comment by patrick-motard** [comment]
@mmatur do you have any idea what could be causing that error now that this PR has been merged?

**Comment by ldez** [comment]
@patrick-motard this issue has been fixed in 1.7

**Comment by patrick-motard** [comment]
@ldez thank you for that information. I upgraded to 1.7 and no longer see errors about unmarshalling. Now I am seeing symptoms identical to #3487.
